### PR TITLE
Enforcing fact relationships

### PR DESCRIPTION
**Pull request by brianedmonds90**
@kyberspache Please review

### CODE REVIEW

File: app/service/planning_svc.py [hunk 1]
```diff
@@ -104,6 +107,34 @@ def __init__(self):
...
+        # get the requirement module from the operation dict
+        for phase in (operation['adversary']['phases']).values():
+            for ab in phase:
+                if ab['id'] == link['ability']:
+                    for req in ab.get('requirements', []):
+                        requirements.append(req)
+        return requirements
+
+    @staticmethod
+    async def _load_requirements(requirements_info):
```
Comment (privateducky) [inline app/service/planning_svc.py]: think you could extract this function to base_service? then this + the parsing (from parsing_svc) + the planner (from operation_svc) could use it.

File: app/service/planning_svc.py [hunk 2]
```diff
@@ -91,10 +92,12 @@ def __init__(self):
                 agent_facts = await self._get_agent_facts(operation['id'], agent['paw'])
                 relevant_facts = await self._build_relevant_facts(variables, operation.get('facts', []), agent_facts)
                 valid_facts = await RuleSet(rules=operation.get('rules', [])).apply_rules(facts=relevant_facts[0])
+                ability_requirements = await self._get_operation_requirements(link, operation)
```
Comment (privateducky) [inline app/service/planning_svc.py]: we should consider calculating ability_requirements outside of the 'for link in links' loop, then we can look up a requirement while we're inside the loop, instead of looping through multiple times.

File: app/service/planning_svc.py (hunk 1 above)
Comment (brianedmonds90) [reply app/service/planning_svc.py]: Sure, I think that's doable

File: app/service/planning_svc.py [hunk 3]
```diff
@@ -104,6 +107,18 @@ def __init__(self):
                 link['command'] = self.encode_string(decoded_test)
         return links
 
+    async def _do_enforcements(self, ability_requirements, operation, link, combo):
+        for requirements_info in ability_requirements:
+            # list of facts that have a relationship
+            requirements_info['fact_relationships'] = [fact for fact in operation['facts'] if
+                                                       len(fact['relationships']) > 0]
```
Comment (brianedmonds90) [inline app/service/planning_svc.py]: I'm going to remove this line. This data is not used

File: app/service/planning_svc.py [hunk 4]
```diff
@@ -104,6 +107,18 @@ def __init__(self):
                 link['command'] = self.encode_string(decoded_test)
         return links
 
+    async def _do_enforcements(self, ability_requirements, operation, link, combo):
+        for requirements_info in ability_requirements:
+            # list of facts that have a relationship
```
Comment (privateducky) [inline app/service/planning_svc.py]: remove the 2 comments in this function. they are repetitive: if a programmer can't figure out what these lines do, these comments are unlikely to help.

File: app/service/base_service.py [hunk 5]
```diff
@@ -79,3 +80,8 @@ class Reason(Enum):
         FACT_DEPENDENCY = 2
         OP_RUNNING = 3
         UNTRUSTED = 4
+
+    @staticmethod
+    async def load_module(module_type, module_info):
```
Comment (privateducky) [inline app/service/base_service.py]: can you change the operation_svc to use this as well? you can do this in a separate PR.

Comment (brianedmonds90) [reply app/service/base_service.py]: Yes I'll do that
